packages/mml-web/src/elements/MaterialManager.ts

```diff
@@ -0,0 +1,203 @@
+
+interface MaterialTextureCacheItem {
+  userMaterials: Map<Material, Material>;
+  texture: THREE.Texture | Promise<THREE.Texture | null>;
+}
+
+interface SharedMaterialItem {
+  material: Material | null;
+  remoteDocument: RemoteDocument | null;
+  userElements: Map<MElement, MElement>;
+}
+
+export class MaterialManager {
+  public mapKeys: (keyof THREE.MeshStandardMaterial)[] = [
+    "map",
+    "lightMap",
+    "aoMap",
+    "emissiveMap",
+    "bumpMap",
+    "normalMap",
+    "roughnessMap",
+    "metalnessMap",
+    "alphaMap",
+    "envMap",
+  ];
+  private static instance: MaterialManager;
+  private textureCache: Map<string, MaterialTextureCacheItem> = new Map();
+  private textureLoader: THREE.TextureLoader;
+  private sharedMaterials: Map<string, SharedMaterialItem> = new Map();
+
+  private constructor() {
+    this.textureLoader = new THREE.TextureLoader();
+  }
+
+  public static getInstance(): MaterialManager {
+    if (!MaterialManager.instance) {
+      MaterialManager.instance = new MaterialManager();
+    }
+    return MaterialManager.instance;
+  }
+
+  static getScopedMaterialKey(remoteAddress: string, materialId: string) {
+    return `${remoteAddress}#${materialId}`;
+  }
+
+  public getTexture(src: string) {
+    return this.textureCache.get(src);
+  }
+
+  public getSharedMaterial(remoteAddress: string, id: string) {
+    const scopedId = MaterialManager.getScopedMaterialKey(remoteAddress, id);
+    return this.sharedMaterials.get(scopedId)?.material;
+  }
+
+  public async loadTexture(src: string, materialElement: Material): Promise<THREE.Texture | null> {
+    if (!src) {
+      return null;
+    }
+    if (this.textureCache.has(src)) {
+      const cacheItem = this.textureCache.get(src)!;
+      cacheItem.userMaterials.set(materialElement, materialElement);
+      return cacheItem.texture;
+    }
+
+    const texturePromise = this.textureLoader.loadAsync(src).catch(() => null);
+    this.textureCache.set(src, {
+      userMaterials: new Map([[materialElement, materialElement]]),
+      texture: texturePromise,
+    });
+    const texture = await texturePromise;
+    if (!texture) {
+      this.textureCache.delete(src);
+      return null;
+    }
+
+    texture.wrapS = THREE.RepeatWrapping;
+    texture.wrapT = THREE.RepeatWrapping;
+    texture.repeat.set(1, 1);
+    return texture;
+  }
+
+  public unloadTexture(src: string, materialElement: Material) {
+    if (this.textureCache.has(src)) {
+      const cacheItem = this.textureCache.get(src)!;
+      cacheItem.userMaterials.delete(materialElement);
+      if (cacheItem.userMaterials.size === 0) {
+        if (cacheItem.texture instanceof THREE.Texture) {
+          cacheItem.texture.dispose();
+        }
+        this.textureCache.delete(src);
+      }
+    }
+  }
+
+  registerSharedMaterial(remoteAddress: string, id: string, material: Material) {
+    const scopedId = MaterialManager.getScopedMaterialKey(remoteAddress, id);
```

Comment:
Using the address of the document avoids _most_ conflicts, but it's not 100% correct because you can include the same document url in multiple `m-frame`s and then the materials collide and it's possible that those two connections to the same address present inconsistent `m-material`s to the client.